WebDriver: [Cocoa] Regression(255359@main) Exception when deleting remote automation session

patrickangle opened the pull request:
#### c46be5c1183bb2076c4996fd4528e6c638fc5a91
<pre>
WebDriver: [Cocoa] Regression(255359@main) Exception when deleting remote automation session
<a href="https://bugs.webkit.org/show_bug.cgi?id=247384">https://bugs.webkit.org/show_bug.cgi?id=247384</a>
rdar://101872145

Reviewed by Wenson Hsieh.

Revert the changes in 255359@main and instead apply a more targeted approach to fixing that specific issue. Previous
attempts to fix this with more state management in `WKWindowVisibilityObserver` created more issues, so a more targeted
fix it is!

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(-[WKWindowVisibilityObserver startObserving:]):
(-[WKWindowVisibilityObserver stopObserving:]):
(WebKit::WebViewImpl::viewWillMoveToWindowImpl):
(WebKit::WebViewImpl::viewWillMoveToWindow):
(WebKit::WebViewImpl::prepareForMoveToWindow):
(-[WKWindowVisibilityObserver _observeWindow:]): Deleted.

Canonical link: <a href="https://commits.webkit.org/256334@main">https://commits.webkit.org/256334@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a3a722e9feafeba03cab3326f8cb997834f7a4aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95443 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4715 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28498 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105025 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4737 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33446 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87813 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100896 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101107 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82058 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30533 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/87273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73373 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/39228 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/36931 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20124 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40912 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42777 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2108 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42901 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39370 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

RemyDemarest commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -260,6 +260,7 @@ - (void)_settingsDidChange:(NSNotification *)notification
 @end
 
 @interface WKWindowVisibilityObserver : NSObject
+@property (nonatomic) BOOL isPendingDuplicateObserverRemoval;
```
```suggestion
@property (nonatomic, getter=isPendingDuplicateObserverRemoval) BOOL pendingDuplicateObserverRemoval;
```

RemyDemarest commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -310,6 +311,9 @@ - (void)startObserving:(NSWindow *)window
         return;
+    
+    ASSERT(!_isObservingWindow);
+    ASSERT(!_isPendingDuplicateObserverRemoval);
```
```suggestion
    ASSERT(!_pendingDuplicateObserverRemoval);
```

RemyDemarest commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -333,15 +337,25 @@ - (void)startObserving:(NSWindow *)window
+    // Clients of `-_prepareForMoveToWindow:completionHandler` will end up invoking `-stopObserving:` twice, once during
+    // preparation and again when actually moving to a new window.
+    if (_isPendingDuplicateObserverRemoval) {
+        _isPendingDuplicateObserverRemoval = NO;
```
```suggestion
    if (_pendingDuplicateObserverRemoval) {
        _pendingDuplicateObserverRemoval = NO;
```

RemyDemarest commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -2408,6 +2407,7 @@ static NSTrackingAreaOptions trackingAreaOptions()
 {
     m_shouldDeferViewInWindowChanges = true;
     viewWillMoveToWindowImpl(targetWindow);
+    [m_windowVisibilityObserver setIsPendingDuplicateObserverRemoval:YES];
```
```suggestion
    m_windowVisibilityObserver.pendingDuplicateObserverRemoval = YES;
```

whsieh replied:
(This one won't quite work, since `m_windowVisibilityObserver` is a RetainPtr; our preference in WebKit is generally to use `setIsPendingDuplicateObserverRemoval:` instead of `.get().pendingDuplicateObserverRemoval`)

whsieh reviewed (approved):
r=mews

patrickangle committed:
WebDriver: [Cocoa] Regression(255359@main) Exception when deleting remote automation session
https://bugs.webkit.org/show_bug.cgi?id=247384
rdar://101872145

Reviewed by Wenson Hsieh.

Revert the changes in 255359@main and instead apply a more targeted approach to fixing that specific issue. Previous
attempts to fix this with more state management in `WKWindowVisibilityObserver` created more issues, so a more targeted
fix it is!

* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(-[WKWindowVisibilityObserver startObserving:]):
(-[WKWindowVisibilityObserver stopObserving:]):
(WebKit::WebViewImpl::viewWillMoveToWindowImpl):
(WebKit::WebViewImpl::viewWillMoveToWindow):
(WebKit::WebViewImpl::prepareForMoveToWindow):
(-[WKWindowVisibilityObserver _observeWindow:]): Deleted.

Canonical link: https://commits.webkit.org/256334@main
